python/ray/includes/object_ref.pxi

```diff
@@ -154,6 +155,9 @@ cdef class ClientObjectRef(ObjectRef):
     def __init__(self, id: Union[bytes, concurrent.futures.Future]):
         self.in_core_worker = False
         self._mutex = threading.Lock()
+        # client worker might be cleaned up before __dealloc__ is called.
```

Comment:
By using weakref instead of an actual ref, is this to avoiding unnecessarily extending the worker's lifetime?

Reply:
No, it's not. I think it's actually shortening the lifetime of the worker.
previous: lifetime = all reference of the worker
now: lifetime = all reference of the worker - reference from ClientObjectRef.

But I think it's ok, since the worker is a "global" context, and if the worker in `ray.util.client` got deleted it means `ClientObjectRef` is no longer valid.

Reply:
Yes, I mean before this change, client object ref does not hold a ref to the worker either. Using weakref instead of ref which is the initial approach in the PR is indeed better.

Reply:
@mwtian  the reason I use weakref is because in `__dealloc__` you won't be able to know whether the underline structure is still alive or not. `__dealloc__` is supposed to release C related things. 

So the purpose of using `weakref` is to check whether the object is still alive or not.